packages/flutter/test/widgets/async_lifecycle_test.dart

```diff
@@ -0,0 +1,64 @@
+import 'package:flutter/widgets.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+class InvalidOnInitLifecycleWidget extends StatefulWidget {
+  const InvalidOnInitLifecycleWidget({Key key}) : super(key: key);
+
+  @override
+  InvalidOnInitLifecycleWidgetState createState() => new InvalidOnInitLifecycleWidgetState();
+}
+
+class InvalidOnInitLifecycleWidgetState extends State<InvalidOnInitLifecycleWidget> {
+  @override
+  void initState() async {
```

Comment:
The semantics of `async` is changing in Dart 2 such that it will become compatible with `initState`. Namely, functions will begin execution synchronously, but complete asynchronously. This restriction might become very inconvenient in the future as there are legitimate use-cases for initiating asynchronous work from within `initState`, such as fire RPC calls. However, we could have a debug `bool` field on the `State` object that flips to `true` when `super.initState` is called. We could then verify that it was flipped synchronously.

Example:

```
// in StatefulElement
_state.initState();
if (!_state._debugSuperInitStateCalled) {
  // throw error
}

// GOOD implementation:
void initState() async {
  super.initState();
  await server.makeRpcCall();
}

// BAD implementation:
void initState() async {
  await server.makeRpcCall();
  super.initState();
}
```

Reply:
There is still plenty of room for errors, even with sync async start.  For example:

```dart
void initState() {
  super.initState();
  // no set state required, since widget hasn't built yet.
  foo = 2;
}

void initState() async {
  super.initState();
  // set state required to work
  foo = await someService();
}
```
Obviously if you don't put the super call first, then you will get a separate error.

Reply:
I think the opportunity for programming errors w.r.t. `setState` is not reduced significantly by banning `async`. The following might look like legit code:

```dart
// Hey, I put assignment inside setState! Why you no work?
setState(() {
  someService.then((result) {
    foo = result;
  });
});
```

This might be a preference thing. Maybe send a PR and see what others think?